### DISCUSSION TITLE
[@sookim4u] 한국 개발자에게 주석이란 어떤 의미일까?

### DISCUSSION THREAD

**Discussion by crowjdh** [Yooii-Studios/Clean-Code#22] [Ideas]
> #21 

S/W 대가가 주석 챕터에서 언급하지 않은 부분이 있다.
그것은 외국인 프로그래머의 경우는 어찌 할 것인가? 
나는 전산과 84학번으로 지금도 현업에서 프로그램을 개발한다.
내가 만나본 대부분의 개발자들은 영어를 잘 못한다. 영어문장으로 된 아래의 경우에서,
한국어로 주석을 달아 놓으면 직관적으로 이해가 쉽다.
책에 주석으로 한국인이 주석을 어찌 대해야 할지 추가해 놓지 않은 점이 아쉽다.

```java
/**
 * @class Existence : 존재에 대하여
 *
 *        데카르트
 *        Cogito, ergo sum (라틴어)
 *        나는 생각한다. 그러므로 나는 존재한다.
 *        I think,therefore I am
 */
```

```java
public class Existence {

    // 1. 보통의 미국인 프로그래머
    // I think, therefore I am --> 동어반복, 무의미한 주석
    void iThinkThereforeIam() {
    }

    /**
     * 2. 라틴어를 좋아하는 미국인 프로그래머
     */
    // I think, therefore I am --> 외국어 이름이어서 유의미한 주석, 그러나 라틴어 매소드 이름은 영어로 바꾸어야 하겠다.
    void cogitoErgoSum() {
    }

    /**
     * 3. 보통의 한국인 프로그래머
     */
    // 나는 생각한다. 그러므로 나는 존재한다. --> 외국어 매소드 이름이어서 유의미한 주석
    void iThinkThereforeIam() {
    }
}
```

**Comment by crowjdh** [comment]
이 경우는 해당 프로젝트를 개발하는 사람들의 주 언어와 관계가 있어 보입니다. 개발자가 전부 한국어를 주로 사용하는(그리고 이후로도 그렇게 될 것이라 생각되는) 경우라면 한글 주석은 `있으나 마나 한 주석`이라기 보다는 `정보를 제공하는 주석`에 가깝지 않은가 생각합니다. 반대로 특정 언어권에 얽메이지 않는 많은 오픈소스 프로젝트 등은 언어에서 오는 문제를 줄이기 위해 (규격화된 용어를 사용하듯이) 영어로 작성되어야 한다고 생각합니다.

다만 코드는 시간이 지남에 따라 항상 바뀌기 때문에, 많은 부분에 번역만을 위한 주석을 다는 것은 `오해할 여지가 있는 주석(혹은 stale한 주석)`이 될 가능성이 높다고 생각합니다. 개인적으로도 주석은 유지보수가 잘 안된다고 생각하기 때문에 주의깊게 읽지 않습니다. 이 부분은 팀 규칙과 PR 리뷰를 통해 개선은 가능하다고 생각하지만 많은 노력과 규율이 필요할 것 같습니다.

**Reply by sookim4u** [reply]
클린코드같은 지침서는 한국의 초보 개발자가 읽어야 하는 내용이고, 처음부터 습관을 잘 들이기에 중요한 기초를 제공한다고 봅니다.
그리고 대부분의 한국 개발자들은 한국의 어떤 기업에서 프로젝트를 하고 있는데, 잘된 영어문장으로 메소드 이름을 짓는것은 거의 영작을 해야 하는 숙제로 보이고, 실제로 영작이 영어에 적합한 단어를 구사하지 못한다는 것을 체험했기에, 좋은 영작으로 함수명을 짓는 것은 좋으나, 콩글리쉬를 이해하는 것은 더욱 어렵다고 봅니다. 그래서 한글 주석을 잘 달아 주어야 하는것이 필요하다는 의미에서 이슈를 띄었고,
유지보수 과정에서 한글 주석을 바꾸지 않는다는 언급을 많이 들었는데, 이는 코딩을 제대로 하지 않는 것과 같은 수준으로 취급해야 한다고 봅니다. 즉 변경이 되면 Code Review과정에서 다시 한번 리팩토링 부분을 생각해 내야 하는데, 이런 환경이 갖춰지지 않는 다고 하면 더 나쁜 콩글리쉬가 양산되어, 코드를 더욱 난해하게 하니, 주석도 함께 수정해야 한다는 원천적인 지적을 합니다.

**Reply by crowjdh** [reply]
영어 작문 능력과 그에 따른 코드 질의 저하에 대해서는 동의합니다. 다만 개인적으로 장기적인 관점에서 영어 능력(읽기, 쓰기, 말하기 불문)은 개발자에게 필수불가결적인 능력이라고 생각합니다.

대다수의 프레임워크/라이브러리의 API 문서, 구글링으로 찾아보는 troubleshooting 들은 대부분 영어로 작성되어 있거나 만약 번역되어 있더라도 최신 버전과 시차를 두고 릴리즈 되는 경우가 많습니다. 그렇기 때문에 이런 문서/글들에 쉽게 접근하기 위해서는 궁극적으로 영어능력의 향상이 필요하다고 생각합니다.
`프로젝트의 완성`이라는 단기적인 목표와 `개발자로써의 역량 배양`이라는 장기적인 목표 사이에서 각자의 상황에 따라 필요한 선택을 해야할 것 같습니다.
